### PR TITLE
Refactor the cache mechanism

### LINKED ISSUE #242: Refactor the cache mechanism
After we implement the support of customized yamls, there are several work items related to the cache mechanism:

1. If multiple customized yamls are specified, but the order is different, how can we save it as the same record.
2. If the links of customized yamls remain the same, are we able to force a refresh of the cache? It is possible that the yaml contents linked by the URLs have been changed.

### PR DESCRIPTION
**Pull request by houshengbo**
Fixes #242

### CODE REVIEW

**Comment by houshengbo** [comment]
cc @Cynocracy

File: pkg/reconciler/common/releases_test.go [hunk 1]
```diff
@@ -980,3 +988,96 @@ func TestTargetManifestPathArray(t *testing.T) {
...
+			Status: v1alpha1.KnativeServingStatus{
+				Version: "0.12.0",
+			},
+		},
+		expectedManifestsPath: "testdata/kodata/empty/empty-resource.yaml",
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			m, _ := InstalledManifest(test.component)
```
Comment (nak3) [inline pkg/reconciler/common/releases_test.go:1063]:
```suggestion
                        m, err := InstalledManifest(test.component)
                        util.AssertEqual(t, err, nil)
```

I think we should verify that error is nil here.

File: pkg/reconciler/common/releases_test.go [hunk 2]
```diff
@@ -980,3 +988,96 @@ func TestTargetManifestPathArray(t *testing.T) {
...
+				Manifests: []string{"testdata/kodata/knative-serving/0.16.1/serving-crd.yaml" + "," +
+					"testdata/kodata/knative-serving/0.16.1/serving-core.yaml",
+					"testdata/kodata/additional-manifests/additional-resource.yaml"},
+			},
+		},
+		expectedManifestsPath: "testdata/kodata/knative-serving/0.16.1/serving-crd.yaml" + "," +
+			"testdata/kodata/knative-serving/0.16.1/serving-core.yaml" + "," +
+			"testdata/kodata/additional-manifests/additional-resource.yaml",
+	}, {
+		name: "knative-serving with status.version unavailable in kodata",
```
Comment (nak3) [inline pkg/reconciler/common/releases_test.go:1047]: Out of curious, this case (Even if we set spec.version to 0.16.1, status becomes 0.12.0) does happen?

Comment (houshengbo) [reply pkg/reconciler/common/releases_test.go:1047]: It is possible.
We have 0.12.0 knative serving installed. After a long time, users installed the latest version of the operator. They changed the spec.version into 0.16.1, and apply the CR. The operator will failed the validation by providing the error message in the CR, that unable to migrate across multiple minor versions.

File: pkg/reconciler/common/releases.go [hunk 3]
```diff
@@ -213,6 +237,20 @@ func FetchManifest(path string) (mf.Manifest, error) {
 	return result, err
 }
 
+// FetchManifestNoCache returns the manifest by reading them from the path, and saves them in the cache.
+func FetchManifestNoCache(path string) (mf.Manifest, error) {
+	result, err := mf.NewManifest(path)
+	if err == nil {
+		cache[path] = result
```
Comment (maximilien) [inline pkg/reconciler/common/releases.go:244]: So if the cache already had results you will overwrite and not return from cache?

Comment (maximilien) [reply pkg/reconciler/common/releases.go:244]: Or maybe you are checking that somewhere else?

File: pkg/reconciler/common/releases_test.go [hunk 4]
```diff
@@ -980,3 +988,97 @@ func TestTargetManifestPathArray(t *testing.T) {
...
+		})
+	}
+}
+
+func TestCache(t *testing.T) {
+	// Make sure to start with empty cache
+	ClearCache()
+	util.AssertEqual(t, len(cache), 0)
+	expectedPath := "testdata/kodata/knative-serving/0.16.1/"
+	manifest, _ := mf.NewManifest(expectedPath)
```
Comment (maximilien) [inline pkg/reconciler/common/releases_test.go:1077]: Why not assert the error here and fail?

File: pkg/reconciler/common/releases_test.go [hunk 5]
```diff
@@ -980,3 +988,97 @@ func TestTargetManifestPathArray(t *testing.T) {
...
+			Status: v1alpha1.KnativeServingStatus{
+				Version: "0.12.0",
+			},
+		},
+		expectedManifestsPath: "testdata/kodata/empty/empty-resource.yaml",
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			m, err := InstalledManifest(test.component)
```
Comment (maximilien) [inline pkg/reconciler/common/releases_test.go:1063]: Fail on `err` here?

**Review by maximilien — CHANGES_REQUESTED** [review]
Left a few comments. Of course so long tests are passing this should be OK since in theory no externally visible behavior should creep out ... :)

File: pkg/reconciler/common/releases.go (hunk 3 above)
Comment (houshengbo) [reply pkg/reconciler/common/releases.go:244]: This function always reads from the path, but save it into the cache.
Operator needs to know what has been installed in the cluster. This is why we still save it into the cache.

File: pkg/reconciler/common/releases_test.go (hunk 5 above)
Comment (houshengbo) [reply pkg/reconciler/common/releases_test.go:1063]: As I add the comment on Line 1064, this function should not raise error.
If the manifest is unavailable, return manifest with no resource.

File: pkg/reconciler/common/releases.go [hunk 6]
```diff
@@ -213,6 +237,20 @@ func FetchManifest(path string) (mf.Manifest, error) {
 	return result, err
 }
 
+// FetchManifestFromPath returns the manifest by reading them from the path, and saves them in the cache.
+func FetchManifestFromPath(path string) (mf.Manifest, error) {
```
Comment (nak3) [inline pkg/reconciler/common/releases.go:241]:
```suggestion
func fetchManifestFromPath(path string) (mf.Manifest, error) {
```

Can you make this function private? It is not called from other packages.

**Review by nak3 — COMMENTED** [review]
LGTM except for one nit.

**Pull request merged**